mountpoint-s3/src/inode.rs

```diff
@@ -597,13 +600,22 @@ impl WriteHandle {
             WriteStatus::LocalOpen => {
                 state.write_status = WriteStatus::Remote;
                 state.stat.size = object_size;
-                match &mut parent_state.kind_data {
-                    InodeKindData::File { .. } => unreachable!("we know parent is a directory"),
-                    InodeKindData::Directory {
-                        children: _,
-                        writing_children,
-                    } => {
-                        writing_children.remove(&inode.ino());
+                remove_from_writing_children(&mut parent_state.kind_data, &inode.ino());
+                if parent_state.write_status != WriteStatus::Remote {
+                    parent_state.write_status = WriteStatus::Remote;
+
+                    let mut ancestor_ino = parent.inner.parent;
+                    let mut child_ino = parent.ino();
+                    loop {
+                        let ancestor = self.inner.get(ancestor_ino)?;
+                        let mut ancestor_state = ancestor.inner.sync.write().unwrap();
```

Comment:
I'm a bit worried here since we're still holding a lock on parent_state and then we try to lock its ancestors. Could deadlock happen from this situation?

Reply:
I think you are right. Updated to lock the ancestors from top to bottom (and removed helper function).